### PR TITLE
Allow setting resource constraints for build

### LINKED ISSUE #10191: Feature: Allow `docker build` to be resource constrained by the caller
Currently the `docker build` process runs under the docker daemon's cgroup, which means it has access to potentially unlimited CPU, Memory, and Disk bandwidth.  We would like to be able to constrain a docker build to a limited set of resources (similar to containers being run via `docker run`).  While today `docker build` currently runs with fairly high privileges, as things like user namespaces are added we will be increasing the security to where ordinary system users will be able to docker build more frequently.  

I can think of two main user scenarios around constraining the resources a build can use:

1. I'm a user on the command line, and my administrator allows me to docker build on top of images that are not running as the root user.  My administrator would like my CPU, memory, and disk usage to be governed by the same limits as my normal user account.
2. I'm hosting docker builds for my company via an API, and I want to limit the resources that any individual build gets so that I can ensure aggressive or malicious builds don't cause the whole server to fall over.

The first use case would be that ideally docker build uses the cgroup the client PID is under to also run the build - resources are then shared between my session and the build, and are accounted for under my cgroup structure.

The second use case is a bit broader - the implementer of the hosting API might want to specify the exact cgroup based on attributes of the user, or simply have all docker builds default to a specific cgroup template.

There might be three increasing complexity features:

1. On daemon startup, provide a flag for a cgroup template (or other resource spec) which all docker builds should be contained under
2. Allow a cgroup or cgroup template (or resource spec) to be provided as an API argument on the build API, to limit the resources available for execution
3. Have the docker build CLI automatically send (or alternatively fetch via the unix socket credentials) the current cgroup of the CLI process (or default resource spec of the user calling the command) and use that as the default value to the daemon.

In our case, we're trying to host docker builds in a reasonably resource constrained fashion, and moreover, to do it from a container (which already has cgroup limits set).  We'd prefer for the build we run to inherit the limits of the caller so that it's easy to reason about the limits of the build (the "builder" container gets X CPU, so the docker build simply shares that).

Do others have a similar need or see the benefit in a feature like this in docker build?  We wanted to get feedback before pulling together a pull request.

### PR DESCRIPTION
**Pull request by cpuguy83**
Closes #10191

Allow `docker build` to be resource constrained,  allowing --cpu-shares, --cpuset-cpus, --memory,
--memory-swap for all containers created by the build job.

### CODE REVIEW

**Comment by icecrime** [comment]
Interesting! Ping @dustinlacewell @dhiltgen you may like this.

**Comment by crosbymichael** [comment]
I'm 100% for this just as long as we make sure these are not commited to the image.

**Comment by cpuguy83** [comment]
Yep, it's all in HostConfig now that resource options have been moved there.

**Comment by cpuguy83** [comment]
So I think we can move this to code review?

File: docs/man/docker-build.1.md [hunk 1]
```diff
@@ -14,6 +14,11 @@ docker-build - Build a new image from the source code at PATH
 [**-q**|**--quiet**[=*false*]]
 [**--rm**[=*true*]]
 [**-t**|**--tag**[=*TAG*]]
+[**-m**|**--memory**[=*MEMORY*]]
+[**--memory-swap**[=*MEMORY-SWAP]]
+[**-c**|**--cpu-shares**[=*0*]]
```
Comment (duglin) [inline docs/man/docker-build.1.md]: any particular reason this one looks like its in italics?

Comment (cpuguy83) [inline docs/man/docker-build.1.md]: hmmm... that looks like a GH rendering issue, looks fine in editor.
This is a copy/paste from `docker-run.md`

Comment (duglin) [inline docs/man/docker-build.1.md]: weird

**Comment by duglin** [comment]
LGTM

**Comment by dustinlacewell** [comment]
Very cool.

**Comment by icecrime** [comment]
LGTM!

**Comment by LK4D4** [comment]
Guess what? Rebase!

**Comment by cpuguy83** [comment]
Rebased.
Always be rebasing!

File: docs/man/docker-build.1.md (hunk 1 above)
Comment (estesp) [inline docs/man/docker-build.1.md]: standard markdown processing of \*\* bold \*\* and \* italics \*.  But in this case it did catch a minor issue as the MEMORY-SWAP param doesn't end with a "`*`", so it continued italics until it processed "termination \*" on the next line.

**Comment by cpuguy83** [comment]
Updated test to show that constraints are not leaked from the build into the image.

**Comment by LK4D4** [comment]
@cpuguy83 Tests looks pretty bad

**Comment by cpuguy83** [comment]
@LK4D4 That's not from this. Run fine locally.

File: docs/man/docker-build.1.md (hunk 1 above)
Comment (duglin) [inline docs/man/docker-build.1.md]: @cpuguy83 can you fix the missing * ?

**Comment by duglin** [comment]
Aside from the * comment, it LGTM me again with the new test

**Comment by cpuguy83** [comment]
@duglin Updated.

**Comment by mrunalp** [comment]
Nice feature. +1

**Comment by cpuguy83** [comment]
cc @moxiegirl @jamtur01 @fredlf @SvenDowideit

**Comment by jamtur01** [comment]
Docs LGTM

**Comment by LK4D4** [comment]
LGTM
And we need one more docs review

File: docs/sources/reference/api/docker_remote_api_v1.18.md [hunk 2]
```diff
@@ -1156,6 +1156,10 @@ Query Parameters:
 -   **pull** - attempt to pull the image even if an older image exists locally
 -   **rm** - remove intermediate containers after a successful build (default behavior)
 -   **forcerm** - always remove intermediate containers (includes rm)
+-   **memory** - set memory limit for build
+-   **memswap** - Total memory (memory + swap), `-1` to disable swap
+-   **cpushares** - CPU shares (relative weight)
+-   **cpusetcpus** - CPUs in which to allow exection, e.g. `0-3`, `-,1`
```
Comment (fredlf) [inline docs/sources/reference/api/docker_remote_api_v1.18.md]: Need a comma after "e.g.,"

**Comment by fredlf** [comment]
One teeny copy-edit nit-pick. Otherwise, docs LGTM.

**Comment by erikh** [comment]
builder changes LGTM

**Comment by cpuguy83** [comment]
@fredlf Updated.

**Comment by fredlf** [comment]
Thanks!

**Comment by cpuguy83** [comment]
Seems like this is ready for merged.

Note, I did have to update the test since the amount of memory allocated for it was just low enough where it seemed to cause the container to fail to run sometimes, and work other times.
The reason it was so low was because json.Unmarshal didn't like decoding anything anything much larger into int64 (as it was converted to float).

**Comment by anandkumarpatel** [comment]
+1 would be a great feature to have

**Pull request merged**